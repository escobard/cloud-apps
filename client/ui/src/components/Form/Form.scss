[id*="-form"]{
  label{
   display: none !important;
  }
  textarea{
    height: 8em !important;
  }
}