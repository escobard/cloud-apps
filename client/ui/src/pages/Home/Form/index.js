export { default } from "pages/Home/Form/Form";
