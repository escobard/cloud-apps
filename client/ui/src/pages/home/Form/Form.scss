[class*="form"]{
  label{
   display: none !important;
  }
  textarea{
    height: 8em !important;
  }
}
