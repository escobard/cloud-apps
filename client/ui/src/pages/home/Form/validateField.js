const validateField = (condition, error) => {
  if (condition) {
    return error;
  }
  return undefined;
};

export default validateField;
