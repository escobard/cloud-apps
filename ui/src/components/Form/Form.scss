#application-form{
  label{
    display: none;
  }
}