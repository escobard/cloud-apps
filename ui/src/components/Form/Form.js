import React, { Fragment, useState } from "react";
import { Button, Form, Message } from "semantic-ui-react";

import "./Form.scss";

/** Dynamically creates a form from the data provided via props.fields
 * @name DynamicForm
 * @dev renders an infinite number of fields based on props.fields, very powerful and expandable
 * @param {string} id, inherited id from parent
 * @param {object[]} fields, contains form field data
 * @param {function} addNote, form submit handler from parent
 * @param {object} message, contains status strings for form message
 * @returns {Component}, DynamicForm
 * */

const DynamicForm = ({ id, fields, addNote, message }) => {
  const formId = `${id}-form`;

  const [formState, setFormState] = useState({});

  /** Submits the form, triggers POST request from parent
   * @name submitForm
   * @dev could be scrapped all together and just use addNote callback, keeping for readability
   * */

  const submitForm = () => {
    const { subject, note } = formState;

    if (addNote) {
      // sends empty strings if undefined to trigger validation
      const subjectValue = subject ? subject.value : "";
      const noteValue = note ? note.value : "";

      addNote(subjectValue, noteValue);
    }
  };

  /** Handles the change of each field's input, when the user types into a field
   * @dev this also creates the input form state for each field
   * @param {string} value, new field value
   * @param {string} fieldKey, state.fieldKey, determines which state to update dynamically
   * */

  const inputChange = (value, fieldKey) => {
    setFormState({
      ...formState,
      [fieldKey]: {
        value
      }
    });
  };

  /** Dynamically renders all fields, based on props.fields
   * @param {object[]} fields, each object contains field name, label, placeholder, error
   * @returns {Component} Form.Input || Form.TextArea
   * */

  const renderFields = fields => {
    return fields.map((field, index) => {
      const { name, label, placeholder } = field;

      // add conditional for input vs textfield

      if (name === "note") {
        return (
          <Form.TextArea
            key={index}
            name={name}
            label={label}
            onChange={e => {
              inputChange(e.target.value, name);
            }}
            placeholder={placeholder}
          />
        );
      }

      return (
        <Form.Input
          key={index}
          name={name}
          label={label}
          onChange={e => {
            inputChange(e.target.value, name);
          }}
          placeholder={placeholder}
        />
      );
    });
  };

  return (
    // TODO - switch to material-ui just for form to match desired look and feel
    <Fragment>
      {fields ? (
        <Form id={formId}>
          {message && (
            <Message
              color={message.status}
              header={message.title}
              content={message.message}
            />
          )}

          {renderFields(fields)}
          <Form.Field onClick={() => submitForm()} control={Button}>
            Add Note
          </Form.Field>
        </Form>
      ) : (
        <p>Form has no input props!</p>
      )}
    </Fragment>
  );
};

export default DynamicForm;
