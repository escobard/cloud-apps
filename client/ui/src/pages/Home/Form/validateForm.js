import validateField from "pages/Home/Form/validateField";

const validateForm = fields => {
  const errors = [];

  fields.map(field => {
    const { condition, error } = field;
    const fieldError = validateField(condition, error);
    return fieldError && errors.push(fieldError);
  });

  return errors;
};

export default validateForm;
