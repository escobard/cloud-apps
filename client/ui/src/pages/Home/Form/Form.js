import React, { useState } from "react";
import { Button, Form as SemanticForm, Message } from "semantic-ui-react";

import PropTypes from "prop-types";

import useAlert from "hooks/useAlert";

import { addNoteFields } from "constants/forms";
import validateForm from "./validateForm";
import "./Form.scss";

const Form = ({ fields, submit }) => {
  const [formState, setFormState] = useState({});
  const { alert, setAlert } = useAlert();

  const submitForm = () => {
    const { subject, note } = formState;

    // sends empty strings if fields are empty to trigger validation
    const subjectValue = subject ? subject.value : "";
    const noteValue = note ? note.value : "";

    const conditions = [
      {
        condition: subjectValue.length < 5,
        error: addNoteFields[0].errors[0]
      },
      {
        condition: noteValue.length < 25,
        error: addNoteFields[1].errors[0]
      }
    ];

    const errors = validateForm(conditions);

    if (errors.length > 0) {
      return setAlert({
        title: "Form error:",
        message: `Form contains the following error(s): ${errors.join(", ")}.`,
        status: "red"
      });
    }
    setAlert({
      title: "Note added!",
      message: "",
      status: "green"
    });

    return submit(subjectValue, noteValue);
  };

  const inputChange = (value, fieldKey) => {
    setFormState(prevState => {
      return {
        ...prevState,
        [fieldKey]: {
          value
        }
      };
    });
  };

  const renderFields = formFields => {
    return formFields.map(field => {
      const { name, label, placeholder } = field;

      if (name === "note") {
        return (
          <SemanticForm.Field key={name + label}>
            <label htmlFor={name}>{label}</label>
            <textarea
              id={name}
              name={name}
              onChange={e => {
                inputChange(e.target.value, name);
              }}
              placeholder={placeholder}
            />
          </SemanticForm.Field>
        );
      }

      return (
        <SemanticForm.Field key={name + label}>
          <label htmlFor={name}>{label}</label>
          <input
            id={name}
            name={name}
            onChange={e => {
              inputChange(e.target.value, name);
            }}
            placeholder={placeholder}
          />
        </SemanticForm.Field>
      );
    });
  };

  return (
    <SemanticForm className="form" onSubmit={() => submitForm()}>
      {alert && alert.status && (
        <Message
          aria-label={alert.status === "red" ? "Alert" : "Update"}
          color={alert.status}
          header={alert.title}
          content={alert.message}
        />
      )}

      {renderFields(fields)}
      <SemanticForm.Field control={Button} aria-label="Submit">
        Add Note
      </SemanticForm.Field>
    </SemanticForm>
  );
};

Form.propTypes = {
  fields: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      placeholder: PropTypes.string.isRequired,
      errors: PropTypes.arrayOf(PropTypes.string).isRequired
    })
  ).isRequired,
  submit: PropTypes.func.isRequired
};

export default Form;
