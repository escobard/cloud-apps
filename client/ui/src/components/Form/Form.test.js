import Form from "./index";
import { addNoteFields } from "Constants";

describe("Form", () => {
  let results;

  const testFunct = (subject, note) => {
    results = { subject, note };
  };

  const props = {
    id: "test-id",
    message: {
      title: "test title",
      message: "test message",
      status: false
    },
    submit: testFunct,
    fields: addNoteFields
  };

  it(">> snapshot is up to date", () => {
    const { container } = render(<Form {...props} />);
    expect(container).toMatchSnapshot();
  });

  it(">> renders essential elements", () => {
    const { getByLabelText } = render(<Form {...props} />);
    expect(getByLabelText("Subject"));
    expect(getByLabelText("Note *"));
    expect(getByLabelText("Submit"));
  });

});
