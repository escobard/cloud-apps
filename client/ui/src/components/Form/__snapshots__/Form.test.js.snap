// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form >> snapshot is up to date 1`] = `
<div>
  <form
    class="ui form form"
  >
    <div
      class="field"
    >
      <label
        for="subject"
      >
        Subject
      </label>
      <input
        id="subject"
        name="subject"
        placeholder="Enter a subject for your note"
      />
    </div>
    <div
      class="field"
    >
      <label
        for="note"
      >
        Note *
      </label>
      <textarea
        id="note"
        name="note"
        placeholder="Enter a description for your note (required)"
      />
    </div>
    <div
      class="field"
    >
      <button
        aria-label="Submit"
        class="ui button"
      >
        Add Note
      </button>
    </div>
  </form>
</div>
`;
